interfaces/many, cmd/snap-confine: miscellaneous policy updates

jdstrand opened the pull request:
- interfaces/many: for interfaces using nameservice abstraction, allow resolved
- interfaces/alsa: allow rw access to /proc/asound
- interfaces/optical-drive: add comment in apparmor policy
- interfaces/camera: allow read access to video4linux sysfs entries
- cmd/snap-confine: silence harmless fsetid denial
- interfaces/network-control: allow reading rfkill sysfs entries and writing to state (LP: #1707612)

The fsetid denial is caused because the newly created /var/lib mimicked directory is created with non-root group (since snap-confine is setuid, not setgid). The chmod() in sc_quirk_create_writable_mimic() causes the denial. The denial is harmless since the capability isn't needed to perform the chmod(), so silence the denial to avoid user confusion since we don't want snap-confine to have the extra permissions.

zyga commented on cmd/snap-confine/snap-confine.apparmor.in:
```diff
@@ -281,6 +281,14 @@
+    # snap-confine sets the directory up correctly, so simply silence the
+    # denial since we don't want to grant the capability as a whole to
+    # snap-confine.
+    deny capability fsetid,
```
Thank you for tracing this down!

jdstrand commented:
Note that finding what caused the denial was tricky because:
* it only happens on classic
* it only happens when the namespace is first being setup
* it only happens when the invoking user is non-root
* there is no EPERM or EACCES in the strace-- the chmod() call works fine
* the kernel doesn't reliably log the denial except after fresh boot

To determine the cause (and recording here for posterity), I:
* rebooted
* added a sleep() call to main() in snap-confine.c
* ran 'hello-world' with the updated snap-confine
* used gdb to attach to snap-confine during the sleep
* 's'tepped, 'c'ontinued (and 'finish'ed) my way through snap-confine until the denial popped out
* added debugging information to the affected function
* rebooted and ran hello-world, which made the kernel log the denial and snap-confine output:
```
sc_quirk_create_writable_mimic(): mimic_dir=/var/lib
sc_quirk_create_writable_mimic(): mimic_dir st_uid=0,st_gid=1000
sc_quirk_create_writable_mimic(): mimic_dir st_mode=41777
sc_quirk_create_writable_mimic(): ref_dir=/snap/core/current//var/lib
sc_quirk_create_writable_mimic(): ref_dir st_uid=0,st_gid=0
sc_quirk_create_writable_mimic(): ref_dir st_mode=40755
sc_quirk_create_writable_mimic() after chown: mimic_dir=/var/lib
sc_quirk_create_writable_mimic() after chown: mimic_dir st_uid=0,st_gid=0
sc_quirk_create_writable_mimic() after chown: mimic_dir st_mode=41777
sc_quirk_create_writable_mimic() after chmod: mimic_dir=/var/lib
sc_quirk_create_writable_mimic() after chmod: mimic_dir st_uid=0,st_gid=0
sc_quirk_create_writable_mimic() after chmod: mimic_dir st_mode=40755
```

Reading `man capabilities` I could tell from the above that we were hitting "set the set-group-ID bit for a file whose GID does not match the filesystem or any of the supplementary GIDs of the calling process." sc_quirk_create_writable_mimic() does the chown before the chmod and changes the group to 'root', but the calling process' gid is still non-root. I think this could maybe be changed to make the chmod call before the chown, but typically you want to chown first so did it this way.

chipaca commented on cmd/snap-confine/snap-confine.apparmor.in:
```diff
@@ -175,7 +175,7 @@
     mount options=(rw private) -> /tmp/snap.rootfs_*/var/lib/snapd/hostfs/,
     pivot_root,
-    # cleanup 
+    # cleanup
```
*giggle* at having to clean up this comment

chipaca commented on cmd/snap-confine/snap-confine.apparmor.in:
```diff
@@ -281,6 +281,14 @@
     mount options=(ro rbind) /snap/{,ubuntu-}core/*/var/lib/** -> /var/lib/**,
     umount /var/lib/snapd/,
     mount options=(move) /tmp/snapd.quirks_*/ -> /var/lib/snapd/,
+    # On classic systems with a setuid root snap-confine when run by non-root
```
what's the rationale for not having a newline before this comment block?

jdstrand replied:
I like to group like rules together with different rule groups separated by spaces (I may not be 100% consistent with this, but it is what I like to do and did here). This rule is about silencing something related to quirks, so I put it there, without a space.

chipaca commented on interfaces/builtin/firewall_control.go:
```diff
@@ -36,6 +36,25 @@ const firewallControlConnectedPlugAppArmor = `
 
 #include <abstractions/nameservice>
 
+# systemd-resolved (not yet included in nameservice abstraction)
```
can't these all be replaced by a suitable `include`?

jdstrand replied:
Upstream decided that while this is suited for the nameservice abstraction, they felt that since non-Ubuntu (and Ubuntu derivatives) don't have dbus mediation, they didn't want to have it in there yet. I did commit it to the Ubuntu archive, but it isn't available anywhere. In cases like this I try to push the change upstream (eg, apparmor, Ubuntu, etc) *and* snapd, then at some future point (series 18?) we can evaluate if we can just use the abstraction.

jdstrand commented:
Have two +1s and the zesty autopkgtest failure is unrelated. Committing.

jdstrand commented:
@mvo5 - can you consider this PR for 2.27 inclusion?

jdstrand commented:
Thanks for the reviews! :)
